crates/noirc_frontend/src/monomorphization/ast.rs

```diff
@@ -223,6 +229,52 @@ impl Type {
     }
 }
 
+pub fn type_of_lvalue(lvalue: &LValue) -> Type {
+    match lvalue {
+        LValue::Ident(ident) => ident.typ.clone(),
+        LValue::Index { element_type, .. } => element_type.clone(),
+        LValue::MemberAccess { object, field_index } => {
+            let tuple_type = type_of_lvalue(object.as_ref());
+            match tuple_type {
+                Type::Tuple(fields) => fields[*field_index].clone(),
+                _ => unreachable!("ICE: Member access on non-tuple type"),
+            }
+        }
+        LValue::Dereference { element_type, .. } => element_type.clone(),
+    }
+}
+
+pub fn type_of(expr: &Expression) -> Type {
+    match expr {
+        Expression::Ident(ident) => ident.typ.clone(),
+        Expression::Literal(lit) => match lit {
+            Literal::Integer(_, typ) => typ.clone(),
+            Literal::Bool(_) => Type::Bool,
+            Literal::Str(str) => Type::String(str.len() as u64),
+            Literal::Array(array) => {
+                Type::Array(array.contents.len() as u64, Box::new(array.element_type.clone()))
+            }
+        },
+        Expression::Block(stmts) => type_of(stmts.last().unwrap()),
+        Expression::Unary(unary) => unary.result_type.clone(),
+        Expression::Binary(_binary) => unreachable!("TODO: How do we get the type of a Binary op"),
+        Expression::Index(index) => index.element_type.clone(),
+        Expression::Cast(cast) => cast.r#type.clone(),
+        Expression::For(_for_expr) => unreachable!("TODO: How do we get the type of a for loop?"),
+        Expression::If(if_expr) => if_expr.typ.clone(),
+        Expression::Tuple(elements) => Type::Tuple(elements.iter().map(type_of).collect()),
+        Expression::ExtractTupleField(tuple, index) => match tuple.as_ref() {
+            Expression::Tuple(fields) => type_of(&fields[*index]),
+            _ => unreachable!("ICE: Tuple field access on non-tuple type"),
+        },
```

Comment:
This may be a non-tuple if it is e.g. a variable: `foo.field` or `foo.0`. If we need the type here we'd need to store it in this variant.

Reply:
type_of removed for now